### PR TITLE
Make `run.ps1` More Resilient To Transient Startup Errors

### PR DESCRIPTION
**Pull request by HarrisonWAffel**
## Problem:
When provisioning windows worker nodes in vSphere, a graceful reboot of the node will result in the Rancher UI showing an error message and declaring the node as failed. This only occurs when restarting the node 'gracefully' via vmware tools. This error is inconsistent, but happens more often than not. 

A few conditions contributed to this issue 

1. Restarting nodes using vmware tools is more CPU intensive than restarting manually using `Restart-Computer`. Additionally, restarting using the vmware tools seems to result in reconfiguration of several network interfaces, which is not seen when restarting manually using `Restart-Computer`. 

2. The `run.ps1` script will stop the RKE2 service and then restart it somewhat quickly, with only about 5-10 seconds between the two operations.

The combination of the above two symptoms can lead to the RKE2 agent failing to create the load balancer due to the required port not being freed in time. The exact error seen is 

`{listen tcp 127.0.0.1:6444: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.}
{Error starting load balancer: listen tcp 127.0.0.1:6444: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.}`

This error is seen as a failure to apply the plan, and a subsequent error in the Rancher UI is displayed. In actuality, the rke2 agent will restart itself after it fails to create the load balancer, and then will then become functional. 

## Solution:

Increase the period of sleep between the stopping of the service and the starting of the service. This ensures enough time is provided to the host so that the port can be freed and used again once the service is restarted. Additionally, attempt to start the RKE2 agent service multiple times, at most 5 times. This ensures that other transient startup errors are handled more appropriately, and we don't report a failing node due to a problem that can be automatically resolved given a small amount of time.

In the event that multiple attempts to start the service _doesn't work_, we will print out the logs of the RKE2 agent service so that support members can relay more accurate debugging information. 

This change impacts the reconnection time of windows nodes slightly, increasing the total time by ~25 seconds. However, the added stability is worth this cost. 

## Testing: 

+ Create a custom RKE2 cluster and register a linux node with the cp/etcd role.
+ Create a windows VM in vSphere, ensure all prerequisites are configured (Containers feature is enabled, elevated powershell access is obatined, VMWare tools are installed properly) 
+ Register the vSphere node as a worker in the custom cluster
+ Wait for the cluster to become available 
+ In the vSphere UI, right click on the VM and, under the Power option, select  `Restart Guest OS`
+ Ensure that the node successfully reconnects to the rancher server and no errors are seen
+ Restart the node several times, ensuring each time a proper reconnection occurs

## QA Considerations 

There are a few of powershell commands that can be used to get rancher-wins and rke2 agent logs, here are some of the ones i've been using. Remember, you need to run these commands in Powershell, and not cmd. On windows server core, you can type `powershell.exe` to get a powershell session. 

#### Get rancher-wins logs 
```powershell
get-winevent -providername rancher-wins | select-object Message | format-table -wrap
```

#### Get RKE2 agent logs 
```powershell
get-eventlog -logname Application -source rke2 | select-object ReplacementStrings | format-table -wrap
```

## Regression Considerations 

None, the node would consistently fail to reconnect prior to this change so trying multiple times shouldn't make it any worse.

### CODE REVIEW

**Comment by brandond** [comment]
@rancher-max would you be able to have someone take a look at this once the March releases are out?

**Comment by Oats87** [comment]
cc: @aiyengar2

**Comment by rancher-max** [comment]
> @rancher-max would you be able to have someone take a look at this once the March releases are out?

Potentially.. we'd probably need some sort of alpha/rc or something cut though to see the changes during provisioning unless Harrison has a viable way for us to test this out without that? 
I'd probably defer to @sowmyav27's help as the hostbosters qa team has more expertise with vsphere+windows, but we should have bandwidth to help after our March patches are out!

File: package/run.ps1
Comment (HarrisonWAffel) [reply package/run.ps1]: This is a good point. While I didn't see any issues with starting the service this way during my testing, I did take some extra time to get a better understanding of how a windows service handles failure conditions to make sure there wouldn't be a problem. I've refactored the PR to rely more on the automatic restarting done by Windows, just to make sure we don't run into anything nasty. 

I've added a larger buffer period after each attempt at starting the service so that windows can try and restart the service itself if possible. I've found that this usually happens after in about one or two tries, which lasts between 30 and 60 seconds.
 
However, in certain cases where a host is under load and RKE2 fails to start, the service may never transition into a 'running' state which results in windows not attempting to automatically restart the service at all. So, I've kept the `for` loop to handle this scenario. I also found that a similar issue can occur with the `Restart-Service` command, wherein it fails to properly process the command and only stops the service.

The Microsoft documentation describes a [scenario where a service will unexpectedly restart if manually restarted and then manually stopped](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-changeserviceconfig2a#remarks), but does not indicate that this will also happen for a service which is running when the 30 second restart period is reached, and I've not seen that behavior when looking through the logs of the service manager itself.

**Comment by HarrisonWAffel** [comment]
I've looking into ways that we can test this without needing an RC or alpha, i'll add a new comment if I can come up with something reasonable for QA

**Comment by HarrisonWAffel** [comment]
I was able to manually create a multi-arch build of `v1.24.11+rke2t1` which includes these changes that could be used for testing pre-rke2 RC. To use it you would just need to set the rancher environment variable `CATTLE_SYSTEM_AGENT_INSTALLER_IMAGE` to `harrisonwaffel/system-agent-installer-`, and then you would be able to provision a `1.24.11` RKE2 cluster on vSphere to test these changes. If QA uses this image to test we can validate the fix before merging this PR, which would be neat, and if any issues are found I can just update this PR.

Let me know what you think, otherwise we would need a new RC of the system-agent-installer since we have to pull the image in one way or another.

**Review by brandond — APPROVED** [review]
LGTM

**Review by jakefhyde — APPROVED** [review]

File: package/run.ps1
Comment (Oats87) [inline package/run.ps1:103]: This is fine as an interim solution but I would love to be able to see a way to validate that the processes have stopped and the ports have freed -- on a super highly loaded system I worry that 30 seconds may not even be enough, I do know below you are retrying startup so there is that, but this is artificially slowing things down potentially.

File: package/run.ps1
Comment (Oats87) [inline package/run.ps1:158]: Nice

File: package/run.ps1
Comment (Oats87) [inline package/run.ps1:195]: Nice

**Review by Oats87 — APPROVED** [review]
LGTM

**Pull request merged**
